packages/performance/src/performance.ts

```diff
@@ -0,0 +1,274 @@
+  readonly supportsResourceEntries = hasGlobal('PerformanceResourceTiming');
+  readonly supportsPaintEntries = hasGlobal('PerformancePaintTiming');
+
+  readonly timeOrigin = referenceTime();
+  readonly supportsDetailedTime = supportsPerformanceObserver;
+  readonly supportsDetailedEvents =
+    this.supportsNavigationEntries &&
+    this.supportsLongtaskEntries &&
+    this.supportsResourceEntries &&
+    this.supportsPaintEntries;
+
+  private inflightNavigation?: InflightNavigation;
+  private navigationTimeout?: ReturnType<typeof setTimeout>;
+  private firstNavigation?: Navigation;
+  private lifecycleEvents: LifecycleEvent[] = [];
+  private navigationCount = 0;
+  private eventHandlers = {
+    navigation: new Set<EventMap['navigation']>(),
+    navigationStart: new Set<EventMap['navigationStart']>(),
+    lifecycleEvent: new Set<EventMap['lifecycleEvent']>(),
+  };
+
+  constructor() {
+    this.start({timeStamp: 0});
+
+    withNavigation(this.start.bind(this));
+
+    if (!this.supportsDetailedTime || !this.supportsNavigationEntries) {
+      withTiming(
+        ({responseStart, domContentLoadedEventStart, loadEventStart}) => {
+          this.lifecycleEvent({
+            type: EventType.TimeToFirstByte,
+            start: responseStart - this.timeOrigin,
+            duration: 0,
+          });
+
+          this.lifecycleEvent({
+            type: EventType.DomContentLoaded,
+            start: domContentLoadedEventStart - this.timeOrigin,
+            duration: 0,
+          });
+
+          this.lifecycleEvent({
+            type: EventType.Load,
+            start: loadEventStart - this.timeOrigin,
+            duration: 0,
+          });
+        },
+      );
+    } else {
+      withEntriesOfType('navigation', entry => {
+        this.lifecycleEvent({
+          type: EventType.TimeToFirstByte,
+          start: entry.responseStart,
+          duration: 0,
+        });
+
+        if (entry.domContentLoadedEventStart > 0) {
+          this.lifecycleEvent({
+            type: EventType.DomContentLoaded,
+            start: entry.domContentLoadedEventStart,
+            duration: 0,
+          });
+        }
+
+        if (entry.loadEventStart > 0) {
+          this.lifecycleEvent({
+            type: EventType.Load,
+            start: entry.loadEventStart,
+            duration: 0,
+          });
+        }
+      });
+    }
+
+    if (this.supportsResourceEntries) {
+      withEntriesOfType('resource', entry => {
+        if (!WATCH_RESOURCE_TYPES.includes(entry.initiatorType)) {
+          return;
+        }
+
+        this.event(
+          {
+            type:
+              entry.initiatorType === 'script'
+                ? EventType.ScriptDownload
+                : EventType.StyleDownload,
+            start: entry.startTime,
+            duration: entry.duration,
+            metadata: {
+              name: entry.name,
+              size: entry.encodedBodySize,
```

Comment:
Sucks that we don't get this info yet ([bah](https://github.com/Shopify/web/issues/9886)).

(I'm just gurning; please keep this in the PR 😛).